Zend/zend_compile.c

```diff
@@ -4712,6 +4712,125 @@ static void zend_compile_ns_call(znode *result, znode *name_node, zend_ast *args
 }
 /* }}} */
 
+static zend_op *zend_compile_rope_add(znode *result, uint32_t num, znode *elem_node);
+static zend_op *zend_compile_rope_add_ex(zend_op *opline, znode *result, uint32_t num, znode *elem_node);
+static void zend_compile_rope_finalize(znode *result, uint32_t j, zend_op *init_opline, zend_op *opline);
+
+static zend_result zend_compile_func_sprintf(znode *result, zend_ast_list *args) /* {{{ */
+{
+	znode format_string;
+
+	if (args->children < 1) {
+		return FAILURE;
+	}
+
+	zend_compile_expr(&format_string, args->child[0]);
+	if (!(format_string.op_type == IS_CONST && Z_TYPE(format_string.u.constant) == IS_STRING)) {
+		zend_do_free(&format_string);
+		return FAILURE;
+	}
+	if (!(Z_STRLEN(format_string.u.constant) < 32)) {
+		zval_ptr_dtor_str(&format_string.u.constant);
+		return FAILURE;
+	}
+
+	char *p;
+	char *end;
+	uint32_t string_placeholder_count;
+	
+	string_placeholder_count = 0;
+	p = Z_STRVAL(format_string.u.constant);
+	end = p + Z_STRLEN(format_string.u.constant);
+	while (p < end) {
+		if (*p != '%') {
+			p++;
+			continue;
+		}
+		char *q = p + 1;
+		if (q == end) {
+			zval_ptr_dtor_str(&format_string.u.constant);
+			return FAILURE;
+		}
+		if (*q == '%') {
+			p++;
+			continue;
+		}
+
+		if (*q != 's') {
+			zval_ptr_dtor_str(&format_string.u.constant);
+			return FAILURE;
+		}
+
+		string_placeholder_count++;
+
+		p = q;
+		p++;
+	}
+
+	if (string_placeholder_count != (args->children - 1)) {
+		zval_ptr_dtor_str(&format_string.u.constant);
+		return FAILURE;
+	}
+
+	uint32_t rope_elements = 0;
+	uint32_t rope_init_lineno = -1;
+	zend_op *opline;
+
+	string_placeholder_count = 0;
+	p = Z_STRVAL(format_string.u.constant);
+	end = p + Z_STRLEN(format_string.u.constant);
+	char *offset = p;
+	while (p < end) {
+		znode elem_node;
+		if (*p != '%') {
+			p++;
+			continue;
+		}
+		char *q = p + 1;
+		ZEND_ASSERT(q < end);
+		if (*q == '%') {
+			p++;
+			continue;
+		}
+		ZEND_ASSERT(*q == 's');
+
+		zend_compile_expr(&elem_node, args->child[1 + string_placeholder_count]);
+		if (rope_elements == 0) {
+			rope_init_lineno = get_next_op_number();
+		}
+		if (p != offset) {
+			znode const_node;
+			const_node.op_type = IS_CONST;
+			ZVAL_STRINGL(&const_node.u.constant, offset, p - offset);
+			opline = zend_compile_rope_add(result, rope_elements++, &const_node);
+		}
+		opline = zend_compile_rope_add(result, rope_elements++, &elem_node);
+
+		string_placeholder_count++;
+
+		p = q;
+		p++;
+		offset = p;
+	}
+	if (p != offset) {
+		if (rope_elements == 0) {
+			rope_init_lineno = get_next_op_number();
+		}
+
+		znode const_node;
+		const_node.op_type = IS_CONST;
+		ZVAL_STRINGL(&const_node.u.constant, offset, p - offset);
+		opline = zend_compile_rope_add(result, rope_elements++, &const_node);
+	}
+
+	zend_op *init_opline = CG(active_op_array)->opcodes + rope_init_lineno;
+	zend_compile_rope_finalize(result, rope_elements, init_opline, opline);
+
+	zval_ptr_dtor_str(&format_string.u.constant);
```

Comment:
This misses the optimizations from `zend_compile_encaps_list()`, which `concat()`s adjacent string constants. I'm not sure how important this is for `sprintf`.

Reply:
Fairly unimportant, I'd say: Using a placeholder to insert a constant string is not really the expected use case of `sprintf()`.

FWIW: Opcache will automatically merge ropes that are completely constant:

```php
<?php

var_dump(sprintf("%s: %s-%s", __FILE__, "foo", "bar"));
```

```
$ sapi/cli/php -d zend_extension=(pwd)/modules/opcache.so -d opcache.enable_cli=1 -d opcache.opt_debug_level=0x10000 test.php

$_main:
     ; (lines=9, args=0, vars=0, tmps=5)
     ; (before optimizer)
     ; php-src/test.php:1-4
     ; return  [] RANGE[0..0]
0000 INIT_FCALL 1 96 string("var_dump")
0001 T1 = ROPE_INIT 5 string("php-src/test.php")
0002 T1 = ROPE_ADD 1 T1 string(": ")
0003 T1 = ROPE_ADD 2 T1 string("foo")
0004 T1 = ROPE_ADD 3 T1 string("-")
0005 T0 = ROPE_END 4 T1 string("bar")
0006 SEND_VAL T0 1
0007 DO_ICALL
0008 RETURN int(1)
LIVE RANGES:
     1: 0001 - 0005 (rope)
string(25) "php-src/test.php: foo-bar"
$ sapi/cli/php -d zend_extension=(pwd)/modules/opcache.so -d opcache.enable_cli=1 -d opcache.opt_debug_level=0x20000 test.php

$_main:
     ; (lines=4, args=0, vars=0, tmps=0)
     ; (after optimizer)
     ; php-src/test.php:1-4
0000 INIT_FCALL 1 96 string("var_dump")
0001 SEND_VAL string("php-src/test.php: foo-bar") 1
0002 DO_ICALL
0003 RETURN int(1)
string(25) "php-src/test.php: foo-bar"
```

Reply:
Fair enough. This can always be added later. And if opcache takes care of it, it's likely fairly unimportant.